File: lib/src/features/foods/data/foods_repository.dart
import 'package:drift/drift.dart';
import 'package:uuid/uuid.dart';

import '../../../core/database/database.dart';
import '../../../core/database/sync.dart';
import 'foods_dao.dart';

/// The offline-first seam. Today it only talks to the local DAO; when the
/// backend arrives, remote read/write and a sync queue live here — the UI
/// above never changes.
class FoodsRepository {
  FoodsRepository(this._dao, {Uuid? uuid}) : _uuid = uuid ?? const Uuid();

  final FoodsDao _dao;
  final Uuid _uuid;

  Stream<List<Food>> watchFoods() => _dao.watchFoods();

  Future<Food?> findById(String id) => _dao.findById(id);

  /// Create (id null) or update an existing food. Stamps sync metadata.
  Future<void> save({
    String? id,
    required String name,
    String? photoPath,
    double carbsGrams = 0,
    double sodiumMg = 0,
    double caffeineMg = 0,
    String? notes,
  }) {
    return _dao.upsert(
      FoodsCompanion(
        id: Value(id ?? _uuid.v4()),
        name: Value(name),
        photoPath: Value(photoPath),
        carbsGrams: Value(carbsGrams),
        sodiumMg: Value(sodiumMg),
        caffeineMg: Value(caffeineMg),
        notes: Value(notes),
        updatedAt: Value(DateTime.now()),
        syncStatus: const Value(SyncStatus.pending),
      ),
    );
  }

  Future<void> delete(String id) => _dao.softDelete(id);
}

File: kmp/composeApp/src/commonMain/kotlin/com/eatrun/features/foods/data/FoodsRepository.kt
package com.eatrun.features.foods.data

import com.eatrun.core.db.FoodDao
import com.eatrun.core.db.FoodEntity
import com.eatrun.core.db.SyncColumns
import kotlinx.coroutines.flow.Flow
import kotlinx.datetime.Clock
import kotlin.uuid.ExperimentalUuidApi
import kotlin.uuid.Uuid

/// The seam the UI talks to. Reads/writes the local DB now; a backend can slip
/// in behind it later without the UI changing (offline-first).
@OptIn(ExperimentalUuidApi::class)
class FoodsRepository(private val dao: FoodDao) {

    fun observeFoods(): Flow<List<FoodEntity>> = dao.observeFoods()

    suspend fun find(id: String): FoodEntity? = dao.findById(id)

    suspend fun softDelete(id: String) {
        dao.softDelete(id, now())
    }

    /// Create (id null) or update a food. On update we keep the existing sync
    /// row but re-stamp it pending, so a later backend re-syncs the change.
    suspend fun save(
        id: String?,
        name: String,
        carbsGrams: Int,
        sodiumMg: Int,
        caffeineMg: Int,
        notes: String,
    ) {
        val existingSync = id?.let { dao.findById(it)?.sync }
        val sync = existingSync?.copy(updatedAt = now(), syncStatus = "pending")
            ?: SyncColumns(id = Uuid.random().toString(), updatedAt = now())
        dao.upsert(
            FoodEntity(
                sync = sync,
                name = name.trim(),
                carbsGrams = carbsGrams,
                sodiumMg = sodiumMg,
                caffeineMg = caffeineMg,
                notes = notes.ifBlank { null },
            ),
        )
    }

    private fun now(): Long = Clock.System.now().toEpochMilliseconds()
}
